### ISSUE TITLE
securityonion-bro-pkg: create a new package for bro-pkg

### ISSUE THREAD

**Issue by dougburks** [Security-Onion-Solutions/security-onion#1324]
Ubuntu 18.04 has a package for `bro-pkg`:
http://manpages.ubuntu.com/manpages/bionic/man1/bro-pkg.1.html

We could build our own bro-pkg based on their packaging.

**Comment by TOoSmOotH** [comment]
Couple thoughts on this... I understand the usefulness of bro-pkg when one has direct internet access. The issue is there are several deployment scenarios that we have that don't have said access. Also this model breaks down when you have large amounts of sensors. In large scale deployments, I have always compiled the plugin or package separately and used salt or the like to distribute it. This gets further complicated IF we put bro in a docker container.

**Comment by JonZeolla** [comment]
@TOoSmOotH  `bro-pkg` should be able to natively handle all of those scenarios.  I've deployed `bro-pkg` in environments that don't have Internet access by setting up a local [Package Source](https://bro-package-manager.readthedocs.io/en/stable/source.html), but they also have [bundles](https://bro-package-manager.readthedocs.io/en/stable/bro-pkg.html#bundle) which are specifically made for deployments without Internet connectivity where you'd like to deploy a package or set of packages.  In your network-restricted large deployment scenario, you could centrally push out a bundle with salt/ansible/puppet/chef/cfengine/... fairly easily, but you could also take the local package source route (personally I've done this in ansible and puppet environments).

Would you mind describing the issue with `bro-pkg` when `bro` is in a docker container further?

**Comment by JonZeolla** [comment]
@dougburks Is there anything I can do that would help with this item?  I'm open to further discussion on the topic of `bro-pkg` not being a fit for certain Security Onion deployment scenarios, but I don't see any roadblocks there yet.

I've done my fair share of deb packaging and related work, and obviously I won't be getting access to do the release myself any time soon, but maybe I can work out the kinks in porting this to 16.04 and include any other standard securityonion customizations you add to packages?  I'm just not sure what those customizations would be.  I could also couple this work with a SO setup PR/wiki documentation on setup and use in the SO ecosystem.

**Comment by dougburks** [comment]
Hi @JonZeolla ,

I appreciate your willingness to help!

I think one concern is that some bro-pkg packages might make changes to Bro log format and thus might still need additional changes when it comes to Logstash parsing and Kibana visualization.  So while bro-pkg might make it easy to load the Bro packages themselves, it still may not be a 100% solution.

Additionally, I think one of the Docker concerns is that containers are ephemeral by design and so if we move Bro into a Docker image and then run bro-pkg, then those Bro packages would only exist for the lifetime of that particular container.

Given some of the concerns above and the fact that we're working on so many other projects right now, we're not able to dedicate any cycles to this at the current time.  But if you would like to work on this, please feel free to do so.  It would need to be a totally self contained package that includes bro-pkg and the Bro source and it would need to be hosted in a Launchpad PPA.

Thanks!

**Comment by JonZeolla** [comment]
I agree, users will need to know what they're installing with bro-pkg and make the related downstream changes, similar to when they manually install/use scripts today.  Many of the packages do not change the log format and instead allow things like adding an additional logging destination or simply identify malicious or unwanted traffic to notice.log.

I'm still not clear on the actual issue with Bro in a Docker image in a world where SO supports bro-pkg.

Right, I understand this is not a priority for you, that's why I'm willing to do the work myself, to whatever extent it's currently possible.  At the very least I'll hack on this during BroCon next month, if not sooner.

**Comment by JonZeolla** [comment]
By the way, I am putting this on my back burner because as of bro 2.6 (not yet released, obviously) the headers are included in the install dir so `bro-pkg` will be able to build binary packages without needing the full source, which makes the process of supporting `bro-pkg` much cleaner.  Once that gets released I'll pick this one up.

**Comment by JonZeolla** [comment]
Depends on #1388

**Comment by dougburks** [comment]
No plans to add this at the current time.
